third_party/gtp_ovs/ovs-gtp-patches/2.15/0024-QFI-support-in-OVS.patch

```diff
@@ -0,0 +1,891 @@
+From 719e7040e76cc230c10fd9b4126802dd465cd827 Mon Sep 17 00:00:00 2001
+From: Prabina Pattnaik <prabinak@wavelabs.ai>
+Date: Mon, 25 Apr 2022 19:15:48 +0000
+Subject: [PATCH 24/24] QFI support in OVS
+
+Signed-off-by: Prabina Pattnaik <prabinak@wavelabs.ai>
+---
+ datapath/linux/compat/gtp.c               | 144 +++++++++++++---------
+ datapath/linux/compat/include/linux/gtp.h |   4 +-
+ include/openvswitch/match.h               |   2 +
+ include/openvswitch/meta-flow.h           |  17 +++
+ include/openvswitch/ofp-actions.h         |  13 ++
+ include/openvswitch/packets.h             |   3 +-
+ lib/flow.c                                |   6 +
+ lib/match.c                               |  17 +++
+ lib/meta-flow.c                           |  19 +++
+ lib/meta-flow.xml                         |   9 ++
+ lib/nx-match.c                            |   2 +-
+ lib/odp-util.c                            |  24 +++-
+ lib/ofp-actions.c                         |  73 +++++++++++
+ lib/packets.h                             |  14 ++-
+ tests/system-layer3-tunnels.at            | 108 ++++++++++++++++
+ tests/tunnel.at                           |  21 ++++
+ 16 files changed, 406 insertions(+), 70 deletions(-)
+
+diff --git a/datapath/linux/compat/gtp.c b/datapath/linux/compat/gtp.c
+index a7033031d..94bc0f7f6 100644
+--- a/datapath/linux/compat/gtp.c
++++ b/datapath/linux/compat/gtp.c
+@@ -113,18 +113,39 @@ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 			udp_tun_rx_dst(skb, sk->sk_family, TUNNEL_KEY, tid, opts_len);
+ #endif
+ 		netdev_dbg(gtp->dev, "attaching metadata_dst to skb, gtp ver %d hdrlen %d\n", gtp_version, hdrlen);
+-		if (unlikely(opts_len)) {
++                if (unlikely(opts_len)) {
++                        struct gtpu_metadata *opts = ip_tunnel_info_opts(&tun_dst->u.tun_info);
++                        struct gtp1_header *gtp1 = (struct gtp1_header *)(skb->data + sizeof(struct udphdr));
++
++                        opts->ver = GTP_METADATA_V1;
++                        opts->flags = gtp1->flags;
++                        opts->type = gtp1->type;
++                        netdev_dbg(gtp->dev, "recved control pkt: flag %x type: %d\n", opts->flags, opts->type);
++                        tun_dst->u.tun_info.key.tun_flags |= TUNNEL_GTPU_OPT;
++                        tun_dst->u.tun_info.options_len = opts_len;
++                        skb->protocol = 0xffff;         // Unknown
++                }
++
++		else {
```

Comment:
if 'opts_len' len is zero, you can not use tunel options, so you need to rework this logic.